pkg/kubelet/dockershim/exec.go

```diff
@@ -110,27 +110,31 @@ func (*NativeExecHandler) ExecInContainer(client libdocker.Interface, container
 	}
 
 	ticker := time.NewTicker(2 * time.Second)
+	execTimeout := time.After(timeout)
 	defer ticker.Stop()
 	count := 0
 	for {
-		inspect, err2 := client.InspectExec(execObj.ID)
-		if err2 != nil {
-			return err2
-		}
-		if !inspect.Running {
-			if inspect.ExitCode != 0 {
-				err = &dockerExitError{inspect}
+		select {
+		case <-execTimeout:
+			return fmt.Errorf("timed out waiting for exec commands %v to finish for container %v", cmd, container.ID)
+		case <-ticker.C:
+			inspect, err2 := client.InspectExec(execObj.ID)
+			if err2 != nil {
+				return err2
+			}
+			if !inspect.Running {
+				if inspect.ExitCode != 0 {
+					err = &dockerExitError{inspect}
+				}
+				break
 			}
-			break
-		}
 
-		count++
-		if count == 5 {
-			klog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
-			break
+			count++
+			if count == 5 {
```

Comment:
I think this count check should be removed now since we're checking the timeout now, but I'll defer to kubelet maintainers since I don't have context on why this is here originally.

Reply:
looks like earlier we just had an infinite loop until `InspectExec` can tell us that the process we started exited... the count was added to break out of it in https://github.com/kubernetes/kubernetes/commit/7748a02e37cb2233cfb59ef6d608457d526d7bf2

Reply:
As I think about this more, I think we should remove it since this means the max timeout you can set for an exec probe with dockershim is 10s (5 x 2s tick).

Reply:
Nvm, I think we need to keep this since it's expected behavior for kubelet to exit at some point with a nil error and allow the exec'd command to  continue running.

Reply:
That or there should be some default timeout here, but don't want to introduce new changes here.